packages/eslint-plugin/typings/eslint-rules.d.ts

```diff
@@ -0,0 +1,12 @@
+declare module 'eslint/lib/rules/*' {
+  import { Node } from 'estree';
+  import { Rule } from 'eslint';
+
+  // intentionally redefined here instead of using Rule.RuleListener
+  // because eslint only uses the `node => void` signature
+  const rule: {
```

Comment:
are you sure we can't use `RuleModule` for `eslint/lib/rules/*`? All `RuleListener` are meant to return void aren't they? (see [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9e8f4db7f300c6b0ef00fece7bfe164d57c79117/types/eslint/index.d.ts#L247))

Reply:
maybe you could use a union type so you don't have to redefine the rest at least.

something like...

```ts
const rule: RuleModule & {
  create(context: Rule.RuleContext): RuleListener & {
    [key: string]: ((node: ESTree.Node) => void)
  }
}
```

example is probably nonsense but you see where im going.

Reply:
The problem is that typescript doesn't have any concept of an override, it just merges the types together.

I.e.
```ts
interface Foo { [k: string]: string }
type Bar = Foo & { [k: string]: number }

// you might think that right now typeof Bar === { [k: string]: number }, but...

const x: Bar = {prop: 1} // Type 'number' is not assignable to type 'string & number'.
```

Because the broken intersection type exists on the `[key: string]` operator, I can't even use a `Diff` type to remove the property from the definition.
Unfortunately I'm pretty certain the only option here is to manually redefine the type.

Reply:
Yeah i see. Last thought is something like:

```ts
const rule: Pick<RuleModule, Exclude<keyof RuleModule, 'create'>> & {
  create(context: Context): {
    [key: string]: ((node: Node) => void);
  }
};
```

Reply:
That'll work, sure!

But at that point, the `Pick<...>` is only copying the definition for the `meta` property, which has the type `meta ?: RuleMetaData`.

However, we know there is always a `meta` property on the imported eslint rules (because they follow good conventions!), so having it as optional will just force us to either do non-null assertions or add in null checks.

Reply:
They do follow good conventions but the type is right, it is optional. So we should be checking anyway wherever we use `meta`. 

A rule can omit it, we shouldn't really be making our own assumptions but rather just following the defined types.